lib/sshutils/scp/scp.go

```diff
@@ -585,28 +662,47 @@ func sendOK(ch io.ReadWriter) error {
 }
 
 type state struct {
-	path     []string
-	finished bool
+	path pathSegments
+	// stat optionally specifies access/modification time for the current file/directory
+	stat *mtimeCmd
 }
 
-func (st *state) push(dir string) {
-	st.path = append(st.path, dir)
+func (r pathSegments) join() string {
+	path := make([]string, 0, len(r))
+	for _, s := range r {
+		path = append(path, s.dir)
+	}
+	return filepath.Join(path...)
 }
 
-func (st *state) pop() error {
-	if st.finished {
-		return trace.Errorf("empty path")
-	}
+var localDir = pathSegments{{dir: "."}}
+
+type pathSegments []pathSegment
+
+type pathSegment struct {
+	dir string
+	// stat optionally specifies access/modification time for the directory
+	stat *mtimeCmd
+}
+
+func (st *state) push(dir string, stat *mtimeCmd) {
+	st.path = append(st.path, pathSegment{dir: dir, stat: stat})
+}
+
+// pop removes the last segment from the current path.
+// Returns the old path as a result
+func (st *state) pop() pathSegments {
 	if len(st.path) == 0 {
-		st.finished = true // allow extra 'E' command in the end
 		return nil
 	}
+	path := st.path
 	st.path = st.path[:len(st.path)-1]
-	return nil
+	st.stat = nil
+	return path
```

Comment:
I don't understand this method. Shouldn't it return `pathSegment` (rather than `pathSegments`)?
Also, it always removes the last element in `st.stat` and then sets that field to nil right after.

Reply:
indeed, I confused `st.path` and `st.stat`